### PR TITLE
Fix CronTrigger.NextFireTime() implementation

### PR DESCRIPTION
**Pull request by joaquinrovira**
Proposed fix for issue #51. I have also updated and validated the tests in [`cron_test.go`](quartz/cron_test.go) with [Cronmaker](http://www.cronmaker.com), an online website that uses the original Quartz library to return the next fire times of a given Cron expression.

Hope this helps, and I welcome any discussion regarding the implementation.

### CODE REVIEW

**Comment by lagzi** [comment]
Hey @reugn, will this be considered soon? I'm currently considering using this wonderful library and this seems like a crucial bug

**Comment by joaquinrovira** [comment]
@reugn, if you have no intention to maintain the project, could you kindly clarify your position? Perhaps we can transfer it to a community-owned repository. It would be a great shame to abandon such a valuable project.

As a sidenote, the cron-related functionality could be moved to a separate library. After a quick search, it appears that the only other project that offers similar functionality is [robfig/cron](https://github.com/robfig/cron), which has been unmaintained since January 2021, with numerous pending pull requests and issues.

File: quartz/csm.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package quartz
+
+import (
+	"time"
+
+	CSM "github.com/reugn/go-quartz/quartz/csm"
+)
+
+func MakeCSMFromFields(prev time.Time, fields []*cronField) *CSM.CronStateMachine {
```
Comment (reugn) [inline quartz/csm.go:9]: Can be private

File: quartz/csm.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+		day = CSM.NewMonthdayNode(prev.Day(), 1, 31, fields[3].values, month, year)
+	}
+	hour := CSM.NewCommonNode(prev.Hour(), 0, 59, fields[2].values)
+	minute := CSM.NewCommonNode(prev.Minute(), 0, 59, fields[1].values)
+	second := CSM.NewCommonNode(prev.Second(), 0, 59, fields[0].values)
+
+	csm := CSM.NewCronStateMachine(second, minute, hour, day, month, year)
+	return csm
+}
+func MakeCSMFromCronTrigger(prev time.Time, trigger *CronTrigger) *CSM.CronStateMachine {
```
Comment (reugn) [inline quartz/csm.go:25]: Can be private

**Review by reugn — COMMENTED** [review]
Hi @joaquinrovira,
Thanks for your time to look into the project and your contribution.
Since the PR is a voluminous rework, I had to postpone reviewing it.
Here are my initial comments before we can merge.
* Please add documentation for all new exported items (possibly make the `csm` package internal)
* Add a `doc.go` to the `csm` package with a brief description of the functionality
* Address linter warnings
* Add an empty line between methods where it is missing

**Comment by joaquinrovira** [comment]
On it! Thank you for your time.

**Comment by joaquinrovira** [comment]
Fix the issues mentioned and added an explanation in `internal/csm/doc.go`.

**Comment by reugn** [comment]
@joaquinrovira, please run `gofmt -w quartz/internal/csm/doc.go`. I  plan to review and merge over the next week.

**Comment by joaquinrovira** [comment]
> @joaquinrovira, please run `gofmt -w quartz/internal/csm/doc.go`. I plan to review and merge over the next week.

Done.

File: quartz/internal/csm/node.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	Value() int
+
+	// Resets the node value to the minimum valid value.
+	Reset()
+
+	// Changes the node value to the next valid value.
+	// Returns true if the value overflowed and false otherwise.
+	Next() bool
+
+	// Check is the current node value is valid.
```
Comment (reugn) [inline quartz/internal/csm/node.go:14]:
```suggestion
	// Checks if the current node value is valid.
```

File: quartz/internal/csm/util.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package csm
+
+// Retruns true is the element is included in the slice.
```
Comment (reugn) [inline quartz/internal/csm/util.go:3]:
```suggestion
// Returns true if the element is included in the slice.
```

**Review by reugn — APPROVED** [review]
Thanks, @joaquinrovira.

**Pull request merged**